File: 2025/09 September/03 Reverse a Doubly Linked List/main.cpp
// Link: https://www.geeksforgeeks.org/problems/reverse-a-doubly-linked-list/1

#include <iostream>
#include <vector>
using namespace std;

class Node
{
public:
    int data;
    Node *next;
    Node *prev;
    Node(int val)
    {
        data = val;
        next = NULL;
        prev = NULL;
    }
};

Node *createLinkedList(vector<int> arr)
{
    Node *head = new Node(arr[0]);
    Node *temp = head;

    for (int i = 1; i < arr.size(); i++)
    {
        Node *newNode = new Node(arr[i]);
        temp->next = newNode;
        newNode->prev = temp;
        temp = temp->next;
    }

    return head;
}

void printLinkedList(Node *head)
{
    Node *temp = head;
    while (temp != NULL)
    {
        cout << temp->data << " ";
        temp = temp->next;
    }
    cout << endl;
}

class Solution
{
public:
    Node *reverse(Node *head)
    {
        if (head == nullptr || head->next == nullptr)
            return head;

        Node *prevNode = NULL;
        Node *currNode = head;

        while (currNode != nullptr)
        {
            prevNode = currNode->prev;
            currNode->prev = currNode->next;
            currNode->next = prevNode;
            currNode = currNode->prev;
        }

        return prevNode->prev;
    }
};

int main()
{
    Solution sol;
    vector<int> arr = {3, 4, 5};
    Node *head = createLinkedList(arr);
    cout << "Original Linked List: " << endl;
    printLinkedList(head);
    head = sol.reverse(head);
    cout << "Reversed Linked List: " << endl;
    printLinkedList(head);

    arr = {75, 122, 59, 196};
    head = createLinkedList(arr);
    cout << "Original Linked List: " << endl;
    printLinkedList(head);
    head = sol.reverse(head);
    cout << "Reversed Linked List: " << endl;
    printLinkedList(head);

    return 0;
}

File: 2025/09 September/03 Reverse a Doubly Linked List/main.py
# Link: https://www.geeksforgeeks.org/problems/reverse-a-doubly-linked-list/1


class Node:
    def __init__(self, val):
        self.data = val
        self.next = None
        self.prev = None


def createLinkedList(arr):
    if not arr:
        return None
    head = Node(arr[0])
    curr = head
    for value in arr[1:]:
        newNode = Node(value)
        curr.next = newNode
        newNode.prev = curr
        curr = newNode
    return head


def printLinkedList(head):
    curr = head
    while curr:
        print(curr.data, end=" ")
        curr = curr.next
    print()


class Solution:
    def reverse(self, head):
        if head is None or head.next is None:
            return head

        prevNode = None
        currNode = head

        while currNode is not None:
            prevNode = currNode.prev
            currNode.prev = currNode.next
            currNode.next = prevNode

            currNode = currNode.prev

        return prevNode.prev


if __name__ == "__main__":
    sol = Solution()
    arr = [3, 4, 5]
    head = createLinkedList(arr)
    print("Original Linked List:")
    printLinkedList(head)
    reversed_head = sol.reverse(head)
    print("Reversed Linked List:")
    printLinkedList(reversed_head)

    arr = [75, 122, 59, 196]
    head = createLinkedList(arr)
    print("Original Linked List:")
    printLinkedList(head)
    reversed_head = sol.reverse(head)
    print("Reversed Linked List:")
    printLinkedList(reversed_head)
